delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/resources/DetectChange.scala

```diff
@@ -0,0 +1,65 @@
+package ch.epfl.bluebrain.nexus.delta.sdk.resources
+
+import cats.effect.IO
+import ch.epfl.bluebrain.nexus.delta.rdf.IriOrBNode.Iri
+import ch.epfl.bluebrain.nexus.delta.rdf.jsonld.CompactedJsonLd
+import ch.epfl.bluebrain.nexus.delta.sdk.jsonld.JsonLdAssembly
+import ch.epfl.bluebrain.nexus.delta.sdk.model.jsonld.RemoteContextRef
+import ch.epfl.bluebrain.nexus.delta.sdk.resources.DetectChange.Current
+import ch.epfl.bluebrain.nexus.delta.sdk.resources.model.ResourceState
+import io.circe.Json
+
+/**
+  * Detect if the new json-ld state introduces changes compared to the current state
+  */
+trait DetectChange {
+  def apply(newValue: JsonLdAssembly, currentState: ResourceState): IO[Boolean] =
+    apply(
+      newValue,
+      Current(currentState.types, currentState.source, currentState.compacted, currentState.remoteContexts)
+    )
+
+  def apply(newValue: JsonLdAssembly, current: Current): IO[Boolean]
+}
+
+object DetectChange {
+
+  final case class Current(
+      types: Set[Iri],
+      source: Json,
+      compacted: CompactedJsonLd,
+      remoteContexts: Set[RemoteContextRef]
+  )
+
+  private val Disabled = new DetectChange {
+
+    override def apply(newValue: JsonLdAssembly, current: Current): IO[Boolean] = IO.pure(true)
+  }
+
+  /**
+    * Default implementation
+    *
+    * There will be a change if:
+    *   - If there is a change in the resource types
+    *   - If there is a change in one of the remote JSON-LD contexts
+    *   - If there is a change in the local JSON-LD context
+    *   - If there is a change in the rest of the payload
+    *
+    * The implementation uses `IO.cede` as comparing source can induce expensive work in the case of large payloads.
+    */
+  private val Impl = new DetectChange {
+
+    override def apply(newValue: JsonLdAssembly, current: Current): IO[Boolean] =
+      IO.cede
+        .as(
+          newValue.types != current.types ||
+            newValue.remoteContexts != current.remoteContexts ||
+            newValue.compacted.ctx != current.compacted.ctx ||
+            newValue.source != current.source
```

Comment:
Enough to guarantee that the expanded for is the same? We did check the graph before. I think it's ok just wondering

Reply:
To me all the cases should be covered:
https://github.com/BlueBrain/nexus/pull/4684/files#diff-43071a23edb35dcb0a07b6c27a989adc39e7a4227b4a70cbb11b845404f89944
To have a change in the resulting expanded value and thus in the graph, there are 3 parameters: the source, the context defined in the resource, the base defined in the project config when the resource does not define one.

1. We check for a change in the source
2.  We check for the local context `.compacted.ctx` that may have been provided by the source or injected if the resource do not define one
3. We check for remote contexts defined in the resource context (We have all contexts here, even the transitive ones).
4. We check for the types, it allows to detect cases where the `@base` is not defined in the local context and so relies on the one defined in the project config for it.